src/unix/udp.c

```diff
@@ -207,13 +268,93 @@ static void uv__udp_recvmsg(uv_udp_t* handle) {
       && handle->recv_cb != NULL);
 }
 
+static void uv__udp_sendmmsg(uv_udp_t* handle) {
+  uv_udp_send_t* req;
+  struct uv__mmsghdr h[MMSG_MAXWIDTH];
+  QUEUE* q;
+  ssize_t npkts;
+  int i;
+
+  if (QUEUE_EMPTY(&handle->write_queue)) {
+    return;
+  }
+
+write_queue_drain:
+  for (i = 0, q = QUEUE_HEAD(&handle->write_queue);
+       i < MMSG_MAXWIDTH && q != &handle->write_queue;
+       ++i, q = QUEUE_HEAD(q)) {
+    assert(q != NULL);
+    req = QUEUE_DATA(q, uv_udp_send_t, queue);
+    assert(req != NULL);
+
+    memset(&h[i], 0, sizeof(h[i]));
+    h[i].msg_hdr.msg_name = &req->addr;
+    h[i].msg_hdr.msg_namelen = (req->addr.ss_family == AF_INET6 ?
+      sizeof(struct sockaddr_in6) : sizeof(struct sockaddr_in));
+    h[i].msg_hdr.msg_iov = (struct iovec*) req->bufs;
+    h[i].msg_hdr.msg_iovlen = req->nbufs;
+  }
+
+  do {
+    npkts = uv__sendmmsg(handle->io_watcher.fd, h, i, 0);
+  } while (npkts == -1 && errno == EINTR);
+
+  if (npkts < 1) {
+    if (errno == EAGAIN || errno == EWOULDBLOCK || errno == ENOBUFS)
+      return;
+    else {
+      for (i = 0, q = QUEUE_HEAD(&handle->write_queue);
+	   i < MMSG_MAXWIDTH && q != &handle->write_queue;
+	   ++i, q = QUEUE_HEAD(q)) {
+        assert(q != NULL);
+        req = QUEUE_DATA(q, uv_udp_send_t, queue);
+        assert(req != NULL);
+
+        req->status = UV__ERR(errno);
+      }
+    }
+    return;
+  }
+  else {
+    for (i = 0, q = QUEUE_HEAD(&handle->write_queue);
+	 i < MMSG_MAXWIDTH && q != &handle->write_queue;
+	 ++i, q = QUEUE_HEAD(&handle->write_queue)) {
+      assert(q != NULL);
+      req = QUEUE_DATA(q, uv_udp_send_t, queue);
+      assert(req != NULL);
+
+      req->status = req->bufs[0].len;
+
+      /* Sending a datagram is an atomic operation: either all data
+       * is written or nothing is (and EMSGSIZE is raised). That is
+       * why we don't handle partial writes. Just pop the request
+       * off the write queue and onto the completed queue, done.
+       */
+      QUEUE_REMOVE(&req->queue);
+      QUEUE_INSERT_TAIL(&handle->write_completed_queue, &req->queue);
+      uv__io_feed(handle->loop, &handle->io_watcher);
+    }
+    /* couldn't batch everything, continue sending (jump to avoid stack growth) */
+    if (!QUEUE_EMPTY(&handle->write_queue))
+      goto write_queue_drain;
+  }
+
+  return;
+}
 
 static void uv__udp_sendmsg(uv_udp_t* handle) {
   uv_udp_send_t* req;
-  QUEUE* q;
   struct msghdr h;
+  QUEUE* q;
   ssize_t size;
 
+  /* TODO: Test of uv_sendmmsg just once */
```

Comment:
Can you add your name to the TODO.

Reply:
I rather wrote the code, not sure if the naming is right:
```
static int uv__recvmmsg_avail = 0;
static int uv__sendmmsg_avail = 0;
```
but naming is the hardest part of CS :).